src/base/snark0.ml

```diff
@@ -65,14 +66,20 @@ struct
       Runner.State.make ~num_inputs ~input ~next_auxiliary ~aux ~system
         ~with_witness:false ()
     in
-    ignore (run t state) ;
+    let state, res = run t state in
+    let res, _ = return_typ.var_to_fields res in
+    let output, _ = return_typ.var_to_fields output in
+    let _state =
+      Array.fold2_exn ~init:state res output ~f:(fun state res output ->
+          fst @@ Checked.run (Checked.assert_equal res output) state)
+    in
```

Comment:
Part of the contract of the `Typ.t`s is that they should ensure that the layout of all values should be identical. This will only throw if they break that contract.

`fst` is a built-in.